File: Tests/SonarScanner.MSBuild.PreProcessor.Test/DataModel/QualityProfile.cs
/*
 * SonarScanner for .NET
 * Copyright (C) 2016-2024 SonarSource SA
 * mailto: info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

using System.Collections.Generic;
using SonarScanner.MSBuild.PreProcessor.Roslyn.Model;

namespace SonarScanner.MSBuild.PreProcessor.Test;

internal class QualityProfile
{
    private readonly ISet<string> projectIds;

    public QualityProfile(string id, string language, string organization)
    {
        Id = id;
        Language = language;
        Organization = organization;
        this.projectIds = new HashSet<string>();
        Rules = new List<SonarRule>();
    }

    public QualityProfile AddProject(string projectKey, string projectBranch = null)
    {
        var projectId = projectKey;
        if (!string.IsNullOrWhiteSpace(projectBranch))
        {
            projectId = projectKey + ":" + projectBranch;
        }

        this.projectIds.Add(projectId);
        return this;
    }

    public QualityProfile AddRule(SonarRule rule)
    {
        Rules.Add(rule);
        return this;
    }

    public string Id { get; }
    public string Language { get; }
    public string Organization { get; }
    public IEnumerable<string> Projects { get { return this.projectIds; } }
    public IList<SonarRule> Rules { get; }
}

File: its/src/test/java/com/sonar/it/scanner/msbuild/utils/QualityProfile.java
/*
 * SonarScanner for .NET
 * Copyright (C) 2016-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.sonar.it.scanner.msbuild.utils;

import java.util.Arrays;
import java.util.List;

// The quality profile files can be found in the /qualityProfiles directory
// To add a new quality profile:
// * Create a file named $ProfileName$.xml in /qualityProfiles
// * Make sure the profile name and the file name match: <profile><name>$ProfileName$</name></profile>
// * Add here: $ProfileName$("$language$");
public enum QualityProfile {
  CS_Empty("cs"),
  CS_S1134("cs"),
  CS_S1134_S2699("cs"),
  CS_S1134_S125("cs"),
  CS_S107("cs"),
  CPP_S106("cs"),
  VB_S3385_S125("vbnet"),
  VB_S3385_S2358("vbnet");

  public final String language;

  QualityProfile(String language) {
    this.language = language;
  }

  public static List<String> allProfiles() {
    return Arrays.stream(QualityProfile.class.getFields()).filter(x -> x.isEnumConstant()).map(x -> x.getName()).toList();
  }
}
